docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+
+- gRPC: A subset of queries may be issued in protocol buffer format to the gRPC
+  interface described above under (4). As noted, this endpoint is deprecated
+  and will be removed in v0.36.
+
+### Opportunities for Simplification
+
+**Claim:** There are too many IPC mechanisms.
+
+The preponderance of ABCI usage is via the Cosmos SDK, which means the
+application and the consensus node are compiled together into a single binary,
+and the consensus node calls the ABCI methods of the application directly as Go
+functions.
+
+TODO: What are the remaining use cases for the socket protocol and gRPC? Ethan
+mentioned some cases like web-based explorer UIs and possibly other tools.
+
+As far as I can tell the primary consumers of the multi-headed "RPC service"
+are the light client and the `tendermint` command-line client. There is
+probably some local use via curl, but I expect that is mostly ad hoc. Ethan
+reports that nodes are often configured with the ports to the RPC service
+blocked, which is good for security but complicates use by the light client.
+
+TODO: Figure out what is going on with the weird pseudo-inheritance service
+interface.  The BaseService type makes sense, but its embeds are more
+complicated.
+
+### Context: Remote Signer Issues
+
+Since the remote signer needs a secure communication channel to exchange keys
+and signatures, and is expected to run truly remotely from the node (i.e., on a
+separate physical server), there is not a whole lot we can do here. We should
+finish the deprecation and removal of the "raw" socket protocol between the
+consensus node and remote signers, but the use of gRPC is appropriate.
+
+The main improvement we can make is to simplify the implementation quite a bit,
+once we no longer need to support both "raw" and gRPC transports.
+
+### Context: ABCI Issues
+
+In the original design of ABCI, the presumption was that all access to the
+application should be mediated by the consensus node. The idea is that outside
+access could change application state and corrupt the consensus process, which
+relies on the application to be deterministic. Of course, even without outside
+access an application could behave nondeterministically, but allowing other
+programs to send it requests was seen as courting trouble.
+
+Conversely, users noted that most of the time, tools written for a particular
+application don't want to talk to the consensus module directly. The
+application "owns" the state machine the consensus engine is replicating, so
+tools that care about application state should talk to the application.
+Otherwise, they would have to bake in knowledge about Tendermint (e.g., its
+interfaces and data structures) just because of the mediation.
+
+For clients to talk directly to the application, however, there is another
+concern: The consensus node is the ABCI _client_, so it is inconvenient for the
+application to "push" work into the consensus module via ABCI itself.  The
```

Comment:
There was a presentation once by Marko (and I've had similar thought as well) of a reverse ABCI. Basically, this is enabling bidirectional communication by having a channel with which Tendermint is the server and the application is the client. This would mean that transactions could be sent directly to Tendermint and not have to be routed around via the RPC (or querying of tendermint specific state for that matter).

Reply:
@marbar3778 is that presentation linked anywhere? I'd be interested to read/watch it (not sure if it's a document or a recording).

I agree that is worth considering. It's a bit outside the scope of this RFC, but I think there's a good case to be made for ABCI to operate between peers rather than client/server (I am choosing my phrasing carefully here: I don't mean "P2P" in the sense of message-passing, I only mean that both endpoints could be equally entitled to send and receive requests from each other)